src/gc-alloc-profiler.cpp

```diff
@@ -0,0 +1,169 @@
+    memcpy(bt_data, static_bt_data, bt_bytes);
+
+    return RawBacktrace{
+        bt_data,
+        bt_size
+    };
+}
+
+// == exported interface ==
+
+JL_DLLEXPORT void jl_start_alloc_profile(int skip_every) {
+    g_alloc_profile = AllocProfile{skip_every};
+
+    for (int i = 0; i < jl_n_threads; i++) {
+        g_alloc_profile.per_thread_profiles.push_back(PerThreadAllocProfile{});
+    }
+
+    g_alloc_profile_enabled = true;
+}
+
+extern "C" {  // Needed since the function doesn't take any arguments.
+
+JL_DLLEXPORT struct RawAllocResults jl_stop_alloc_profile() {
+    g_alloc_profile_enabled = false;
+
+    // combine allocs
+    // TODO: interleave to preserve ordering
+    for (const auto& profile : g_alloc_profile.per_thread_profiles) {
+        for (const auto& alloc : profile.allocs) {
+            g_combined_results.combined_allocs.push_back(alloc);
+        }
+    }
+
+    // package up frees
+    for (const auto& profile : g_alloc_profile.per_thread_profiles) {
+        for (const auto& free_info : profile.frees_by_type_address) {
+            g_combined_results.combined_frees.push_back(FreeInfo{
+                free_info.first,
+                free_info.second
+            });
+        }
+    }
+
+    return RawAllocResults{
+        g_combined_results.combined_allocs.data(),
+        g_combined_results.combined_allocs.size(),
+        g_combined_results.combined_frees.data(),
+        g_combined_results.combined_frees.size()
+    };
+}
+
+JL_DLLEXPORT void jl_free_alloc_profile() {
+    for (auto profile : g_alloc_profile.per_thread_profiles) {
+        for (auto alloc : profile.allocs) {
+            free(alloc.backtrace.data);
+        }
+    }
+    g_alloc_profile.per_thread_profiles.clear();
+
+    g_combined_results.combined_allocs.clear();
+    g_combined_results.combined_frees.clear();
+}
+
+}
+
+// == callbacks called into by the outside ==
+
+void _record_allocated_value(jl_value_t *val, size_t size) JL_NOTSAFEPOINT {
+    auto& global_profile = g_alloc_profile;
+
+    auto& profile = global_profile.per_thread_profiles[jl_threadid()];
+
+    profile.alloc_counter++;
+    auto diff = profile.alloc_counter - profile.last_recorded_alloc;
+    if (diff < g_alloc_profile.skip_every) {
+        return;
+    }
+    profile.last_recorded_alloc = profile.alloc_counter;
+
+    auto type = (jl_datatype_t*)jl_typeof(val);
+
+    profile.type_address_by_value_address[(size_t)val] = (size_t)type;
+
+    profile.allocs.emplace_back(RawAlloc{
+        type,
+        get_raw_backtrace(),
+        size
+    });
+}
+
+void _record_freed_value(jl_taggedvalue_t *tagged_val) JL_NOTSAFEPOINT {
+    jl_value_t *val = jl_valueof(tagged_val);
+
+    auto& profile = g_alloc_profile.per_thread_profiles[jl_threadid()];
+
+    auto value_address = (size_t)val;
+    auto type_address = profile.type_address_by_value_address.find(value_address);
+    if (type_address == profile.type_address_by_value_address.end()) {
+        return; // TODO: warn
```

Comment:
Maybe we can emit an "unknown" type here? -- like `nullptr` or some other global const sentinal value or something, and handle it on the julia side?

Reply:
removed this when removing the garbage profiler